File: code_provenance/reports/terminal_report.py
"""Terminal report formatter using rich for colored output."""

from __future__ import annotations

from code_provenance.types import LineRange, ScanResult

# Color codes for terminal output (ANSI)
RED = "\033[31m"
GREEN = "\033[32m"
YELLOW = "\033[33m"
BOLD = "\033[1m"
DIM = "\033[2m"
RESET = "\033[0m"

ICONS = {
    "ai-generated": "\U0001f916",  # robot
    "human-written": "\U0001f464",  # bust in silhouette
    "unknown": "\u2753",  # question mark
}


def _classification_color(classification: str) -> str:
    """Get ANSI color code for a classification."""
    if classification == "ai-generated":
        return RED
    if classification == "human-written":
        return GREEN
    return YELLOW


def _confidence_color(level: str) -> str:
    """Get ANSI color code for a confidence level."""
    if level == "HIGH":
        return GREEN
    if level == "MEDIUM":
        return YELLOW
    return RED


def _range_detail(line_range: LineRange) -> str:
    """Generate descriptive detail for a line range."""
    if line_range.classification.value == "ai-generated":
        if line_range.model_attribution:
            model = line_range.model_attribution.model.value
            name = "AI" if model == "unknown" else model.capitalize()
            return f"{name}-style patterns"
        return "AI patterns detected"
    if line_range.classification.value == "human-written":
        # Pick the most descriptive signal
        naming = next((s for s in line_range.signals if s.detector == "naming-patterns"), None)
        if naming and naming.strength < 0.3:
            return "Domain-specific naming"
        entropy = next((s for s in line_range.signals if s.detector == "entropy"), None)
        if entropy and entropy.strength < 0.3:
            return "Irregular style, high entropy"
        return "Human patterns detected"
    return "Insufficient signals"


def _format_label(classification: str) -> str:
    """Format classification label for display."""
    if classification == "ai-generated":
        return "AI-generated"
    if classification == "human-written":
        return "Human-written"
    return "Unknown"


class TerminalFormatter:
    """Format scan results for terminal display with colors."""

    def format(self, result: ScanResult) -> str:
        file = result.file
        ranges = result.ranges
        summary = result.summary
        lines: list[str] = []

        # Header
        lines.append(f"{BOLD}Code Provenance v{result.metadata.tool_version}{RESET}")
        lines.append(f"{DIM}{chr(0x2500) * 30}{RESET}")
        lines.append("")

        # Empty file
        if file.total_lines == 0:
            lines.append(f"{YELLOW}No code to analyze (empty file){RESET}")
            return "\n".join(lines)

        # File info
        parser_note = f"{DIM} (generic parser){RESET}" if file.parser_used == "generic" else ""
        lines.append(
            f"\U0001f4ca {BOLD}{file.relative_path}{RESET} "
            f"({file.total_lines} lines){parser_note}"
        )
        lines.append("")

        # Line ranges
        for r in ranges:
            icon = ICONS.get(r.classification.value, "\u2753")
            label = _format_label(r.classification.value)
            color = _classification_color(r.classification.value)
            detail = _range_detail(r)

            line_range_str = f"Lines {r.start_line}-{r.end_line}:"
            padded = line_range_str.ljust(18)

            lines.append(
                f"{padded}{icon} {color}{label}{RESET}  ({r.confidence}%)  {DIM}{detail}{RESET}"
            )

        # Summary
        lines.append("")
        ai_pct = summary.ai_percentage
        human_pct = summary.human_percentage
        ai_part = (
            f"{RED}{ai_pct}% AI-generated{RESET}"
            if ai_pct > 0
            else f"{ai_pct}% AI-generated"
        )
        human_part = (
            f"{GREEN}{human_pct}% Human-written{RESET}"
            if human_pct > 0
            else f"{human_pct}% Human-written"
        )
        lines.append(f"Summary: {ai_part} | {human_part}")

        conf_color = _confidence_color(summary.overall_confidence.value)
        lines.append(f"Confidence: {conf_color}{summary.overall_confidence.value}{RESET}")

        return "\n".join(lines)


terminal_formatter = TerminalFormatter()

File: src/reports/terminal-report.ts
import kleur from 'kleur';
import type { ReportFormatter, ScanResult, LineRange } from '../types.js';

const ICONS = {
  'ai-generated': '🤖',
  'human-written': '👤',
  'unknown': '❓',
} as const;

function classificationColor(classification: string): (text: string) => string {
  switch (classification) {
    case 'ai-generated': return kleur.red;
    case 'human-written': return kleur.green;
    default: return kleur.yellow;
  }
}

function confidenceColor(level: string): (text: string) => string {
  switch (level) {
    case 'HIGH': return kleur.green;
    case 'MEDIUM': return kleur.yellow;
    default: return kleur.red;
  }
}

function rangeDetail(range: LineRange): string {
  if (range.classification === 'ai-generated') {
    if (range.modelAttribution) {
      const model = range.modelAttribution.model;
      const name = model === 'unknown' ? 'AI' : model.charAt(0).toUpperCase() + model.slice(1);
      return `${name}-style patterns`;
    }
    return 'AI patterns detected';
  }
  if (range.classification === 'human-written') {
    // Pick the most descriptive signal
    const naming = range.signals.find((s) => s.detector === 'naming-patterns');
    if (naming && naming.strength < 0.3) return 'Domain-specific naming';
    const entropy = range.signals.find((s) => s.detector === 'entropy');
    if (entropy && entropy.strength < 0.3) return 'Irregular style, high entropy';
    return 'Human patterns detected';
  }
  return 'Insufficient signals';
}

function formatLabel(classification: string): string {
  switch (classification) {
    case 'ai-generated': return 'AI-generated';
    case 'human-written': return 'Human-written';
    default: return 'Unknown';
  }
}

export const terminalFormatter: ReportFormatter = {
  format(result: ScanResult): string {
    const { file, ranges, summary } = result;
    const lines: string[] = [];

    // Header
    lines.push(kleur.bold(`Code Provenance v${result.metadata.toolVersion}`));
    lines.push(kleur.dim('─'.repeat(30)));
    lines.push('');

    // Empty file
    if (file.totalLines === 0) {
      lines.push(kleur.yellow('No code to analyze (empty file)'));
      return lines.join('\n');
    }

    // File info
    const parserNote = file.parserUsed === 'generic'
      ? kleur.dim(' (generic parser)')
      : '';
    lines.push(`📊 ${kleur.bold(file.relativePath)} (${file.totalLines} lines)${parserNote}`);
    lines.push('');

    // Line ranges
    for (const range of ranges) {
      const icon = ICONS[range.classification];
      const label = formatLabel(range.classification);
      const colorFn = classificationColor(range.classification);
      const detail = rangeDetail(range);

      const lineRange = `Lines ${range.startLine}-${range.endLine}:`;
      const padded = lineRange.padEnd(18);

      lines.push(
        `${padded}${icon} ${colorFn(label)}  (${range.confidence}%)  ${kleur.dim(detail)}`,
      );
    }

    // Summary
    lines.push('');
    const aiPart = summary.aiPercentage > 0
      ? kleur.red(`${summary.aiPercentage}% AI-generated`)
      : `${summary.aiPercentage}% AI-generated`;
    const humanPart = summary.humanPercentage > 0
      ? kleur.green(`${summary.humanPercentage}% Human-written`)
      : `${summary.humanPercentage}% Human-written`;
    lines.push(`Summary: ${aiPart} | ${humanPart}`);

    const confColor = confidenceColor(summary.overallConfidence);
    lines.push(`Confidence: ${confColor(summary.overallConfidence)}`);

    return lines.join('\n');
  },
};
